source/ftrack_connect_pipeline_qt/ui/assembler/base/__init__.py

```diff
@@ -363,41 +374,51 @@ def extract_components(self, versions):
                         if set(file_formats).intersection(
                             set([component_extension])
                         ):
-                            # Construct fragment
-                            definition_fragment = {
-                                'components': [copy.deepcopy(d_component)]
-                            }
-                            definition_fragment['components'][0][
-                                'name'
-                            ] = component_name_effective
+                            # Construct definition fragment
+                            definition_fragment = DefinitionObject({})
                             for key in definition:
-                                if key not in [
-                                    'components',
-                                ]:
+                                if key == core_constants.COMPONENTS:
+                                    definition_fragment[key] = DefinitionList(
+                                        [
+                                            DefinitionObject(
+                                                d_component.to_dict()
+                                            )
+                                        ]
+                                    )
+                                    definition_fragment.get_first(
+                                        type=core_constants.COMPONENT
+                                    )['name'] = component_name_effective
+                                else:
+                                    # Copy the category
```

Comment:
No need to do this. deepcopy is working in the definitionObject as well so the old code should be working as it's simpliear and easier to understand, isn't it?

Reply:
The old code put components before contexts, thus making the factory render the progress widget in wrong order. Keeping the order by itering over the keys as the appear in the definition to copy from.